### DISCUSSION TITLE
Upload from a node backend

### DISCUSSION THREAD

**Discussion by gautamsi** [urql-graphql/urql#1995] [Q&A] [answered]
I am trying to build a migration scenario which involves uploading file. I am unable to figure out how to get upload working with `multipartFetchExchange` in node backend.

setup:

```ts
import * as fetch from 'node-fetch';
import { createClient, gql } from '@urql/core';

const client = createClient({ url: 'http://localhost:3002/api/graphql', fetch: fetch, exchanges: [multipartFetchExchange] });
```

when I try this 
```ts
    const filePath = path.resolve(`${__dirname}/test.jpg`);
    const result = await client.mutation(gql`mutation updateUser($id: ID!, $data: UserUpdateInput){  updateUser(id:$id, data:$data) { id profilePhoto { src}}}`, {
      id: "asjhdgajhdgas",
      data: {
        profilePhoto: {
           upload: {
             createReadStream: () => fs.createReadStream(filePath),
             filename: path.basename(filePath),
             // @ts-ignore
             mimetype: mime.getType(filePath),
             encoding: 'utf-8',
           }
         }
      }
    }).toPromise();
```

any pointers? 

what I understood from code is that `multipartFetchExchange` uses `extract-files` which is for browser based environment not Nodejs

**Comment by JoviDeCroock** [comment, marked as answer]
You are correct `extract-files` is tailored towards web and react-native targets rather than Node, I wonder if it'd be possible to use the `File` constructor that they look for in this package https://github.com/jaydenseric/extract-files/blob/6d8f725e9b9973fd564bfdd784cd9b2ff58f8470/public/isExtractableFile.js#L31

EDIT: doesn't seem Node supports `Blob` and `File` ootb (only in Node 16 https://nodejs.org/dist/latest-v16.x/docs/api/buffer.html#buffer_class_blob)

**Reply by gautamsi** [reply]
after hours of investigation I was able to copy the implementation of `multipartFetchExchange` and change few lines

for anyone interested

copy the file `https://github.com/FormidableLabs/urql/blob/main/exchanges/multipart-fetch/src/multipartFetchExchange.ts` to your repo

1. add following imports
 ```ts
import fs from 'fs';
import FormData from 'form-data';
 ```
2. change how the `extractFiles` was called with this
```ts
const { files, clone: variables } = extractFiles({
...operation.variables,
},'', value => value instanceof fs.ReadStream);
```
this is basically using the `isExtractableFile` argument along with path argument `,'', value => value instanceof fs.ReadStream`

3. import exchange from your file rather than the official package.

this may not be perfect but worked for me

**Reply by gautamsi** [reply]
thanks @JoviDeCroock for the reply
